src/controllers/VueController.php

```diff
@@ -117,23 +84,31 @@ protected static function convertCurrency(float $amount, string $currency, bool
      */
     private function fetchOrders($id = null) : array
     {
-        $single = $_GET['purchasableId'] ?? $id;
-        $start  = $this->start_date ?: '1970-01-01 00:00:00';
-        $end    = $this->end_date ?: gmdate('Y-m-d 23:59:59');
-        $orders = Order::find()->dateOrdered(['and', ">= {$start}", "< {$end}"]);
-        $url    = $_SERVER['REQUEST_URI'];
-        $path   = basename(parse_url($url, PHP_URL_PATH));
+        $single       = $_GET['purchasableId'] ?? $id;
+        $currentStart = \DateTime::createFromFormat('Y-m-d H:i:s', $this->start_date)->format('Y-m-d 00:00:00');
+        $start        = \DateTime::createFromFormat('Y-m-d H:i:s', $this->start_date);
+        $end          = \DateTime::createFromFormat('Y-m-d H:i:s', $this->end_date);
+        // nuber of days in selected range
+        $numDays  = $end->diff($start)->format("%r%a");
+        // get the new start date based on what the previous period would be
+        $newStart = $start->modify($numDays . ' day')->format('Y-m-d 00:00:00');
+        $newEnd   = $end->format('Y-m-d 23:59:59');
+        // query the previous perior and selected range based on new start date
+        $orders   = Order::find()->dateOrdered(['and', ">= {$newStart}", "< {$newEnd}"])->distinct()->orderBy('dateOrdered desc');
+        $url      = $_SERVER['REQUEST_URI'];
+        $path     = basename(parse_url($url, PHP_URL_PATH));
+        $result   = [];
 
         if ($single) {
             $single = Variant::find()->id($single)->one();
             $orders->hasPurchasables([$single]);
-        }
-
-        if ($single || $path === 'sales') {
             $orders->orderStatusId('< 4');
         }
 
-        return $orders->distinct()->orderBy('dateOrdered desc')->all() ?: [];
+        $result['previousPeriod'] = $orders->dateOrdered(['and', ">= {$newStart}", "< {$currentStart}"])->all();
+        $result['currentPeriod']  = $orders->dateOrdered(['and', ">= {$currentStart}", "< {$newEnd}"])->all();
```

Comment:
It's only returning the selected period. Previous one is only used internally for stats.

Reply:
If the get-orders action only returns the current period, then should the fetchOrders function still be doing the extra work?

If the get-stats action returns stats for both periods, it could make 2 separate calls to fetchOrders:

```diff
    private function _getStats()
    {
-        $orders            = $this->fetchOrders();
-        $previousOrders    = $orders['previousPeriod'];
-        $currentOrders     = $orders['currentPeriod'];
+        $previousOrders    = $this->fetchOrders($previousPeriodStart, $previousPeriodEnd);
+        $currentOrders     = $this->fetchOrders($currentPeriodStart, $currentPeriodEnd); 
        $numPreviousOrders = count($previousOrders);
        $numCurrentOrders  = count($currentOrders);

```

---

There's no reason to set `$this->orders` here: 

https://github.com/FosterCommerce/craft-commerceinsights/blob/98d98fc66609132105f66e738abd0059712cc86d/src/controllers/VueController.php#L48-L52

The _getOrders function makes a call to fetchOrders already, so we'll be running 4 queries to fetch a the dataset for the current period.

In fact, I don't think `$this->orders` is used anywhere in the controller.

Reply:
2 queries is less efficient than one because they can't be run in parallel. Since there won't be separate `get-stats` and `get-orders` endpoints in the future, instead only 1 API call, I'd rather keep this as-is. Removed `$this->orders`.